design-documents/graph-ql/coverage/Cart.graphqls

```diff
@@ -11,22 +11,134 @@ type CartQueryOutput {
 }
 
 type Cart {
+    id: ID! @doc(description: "The ID of the cart.")
+    items: [CartItemInterface]
+    applied_coupons: [AppliedCoupon] @doc(description:"An array of `AppliedCoupon` objects. Each object contains the `code` text attribute, which specifies the coupon code")
+    email: String
+    shipping_addresses: [ShippingCartAddress]!
+    billing_address: BillingCartAddress
+    available_payment_methods: [AvailablePaymentMethod] @doc(description: "Available payment methods")
+    selected_payment_method: SelectedPaymentMethod
+    prices: CartPrices
+    total_quantity: Float!
+    is_virtual: Boolean!
+}
+
+type AvailablePaymentMethod {
+    code: String! @doc(description: "The payment method code")
+    title: String! @doc(description: "The payment method title.")
+}
+
+type SelectedPaymentMethod {
+    code: String! @doc(description: "The payment method code")
+    title: String! @doc(description: "The payment method title.")
+    purchase_order_number: String @doc(description: "The purchase order number.")
+}
+
+interface CartItemInterface @typeResolver(class: "Magento\\QuoteGraphQl\\Model\\Resolver\\CartItemTypeResolver") {
     id: String!
+    quantity: Float!
+    prices: CartItemPrices
+    product: ProductInterface!
+    customizable_options: [SelectedCustomizableOption]
+}
+
+type SimpleCartItem implements CartItemInterface @doc(description: "Simple Cart Item") {
+}
 
-    line_items_count: Int!
-    items_quantity: Float!
+type VirtualCartItem implements CartItemInterface @doc(description: "Virtual Cart Item") {
+}
 
-    selected_payment_method: CheckoutPaymentMethod
-    available_payment_methods: [CheckoutPaymentMethod]!
+type ConfigurableCartItem implements CartItemInterface {
+    child_sku: String! @doc(description: "SKU of the simple product corresponding to a set of selected configurable options.")
+    configurable_options: [SelectedConfigurableOption!]
+}
 
-    customer: CheckoutCustomer
-    customer_notes: String
+type DownloadableCartItem implements CartItemInterface @doc(description: "Downloadable Cart Item") {
+    links: [DownloadableProductLinks] @deprecated(description: "Type was renamed from plural to singular, also link now has different ID type")
+    links_v2: [DownloadableProductLink] @doc(description: "An array containing information about the selected links")
+    samples: [DownloadableProductSamples] @doc(description: "DownloadableProductSamples defines characteristics of a downloadable product")
+}
 
-    gift_cards_amount_used: Money
-    applied_gift_cards: [CartGiftCard]
+type BundleCartItem implements CartItemInterface {
+    bundle_options: [SelectedBundleOption!]
+}
 
-    is_multishipping: Boolean!
-    is_virtual: Boolean!
+type GiftCardCartItem implements CartItemInterface {
+    sender_name: String!
```

Comment:
There might be an issue from the implementation perspective. Since this type implement interface with type resolver, most likely we will not be able to create resolver for `GiftCardCartItem` to return all flat values like `sender_name`, `recepient_name` and `message`. As a result, we might need to create a separate resolver for every mentioned field. This fact might be an issue from the performance perspective. 

Maybe we should consider introducing some intermediate entity like

```graphql
type GiftCardCartItemOptions (@resolver ...) {
  sender_name: String!
  recepient_name: String!
  message: String
}

...

type GiftCardCartItem implements CartItemInterface { 
  gift_cart_options: GiftCardCartItemoptions
  amount: SelectedGiftCardAmount
}
```

In that way, we may create just one resolver for `GiftCardCartItemoptions` instead of resolving every field separately. What do you think? 

Thank you.

Reply:
I would really prefer to keep schema-first approach. How about introducing some data provider with internal state and reusing it from all resolvers to improve performance?

Reply:
Discussed with the team working on the implementation, agreed not to introduce wrapper.